ddsrecorder_participants/src/cpp/recorder/mcap/McapHandler.cpp

```diff
@@ -114,160 +115,225 @@ McapHandler::~McapHandler()
-    }
 
-    if (received_types_.count(topic.type_name) != 0)
-    {
-        // Schema available -> add to buffer
-        add_data_nts_(msg, topic);
-    }
-    else
-    {
-        if (state_ == McapHandlerStateCode::RUNNING)
+        if (received_types_.count(topic.type_name) != 0)
         {
-            if (configuration_.max_pending_samples == 0)
+            // Schema available -> add to buffer
+            // Check if there is enough space available before adding the message to buffer
+            mcap_size_tracker_.check_and_increase_mcap_size(msg.dataSize);
+            add_data_nts_(msg, topic);
+        }
+        else
+        {
+            if (state_ == McapHandlerStateCode::RUNNING)
             {
-                if (configuration_.only_with_schema)
+                if (configuration_.max_pending_samples == 0)
                 {
-                    // No schema available + no pending samples + only_with_schema -> Discard message
-                    return;
+                    if (configuration_.only_with_schema)
+                    {
+                        // No schema available + no pending samples + only_with_schema -> Discard message
+                        return;
+                    }
+                    else
+                    {
+                        // No schema available + no pending samples -> Add to buffer with blank schema
+                        mcap_size_tracker_.check_and_increase_mcap_size(msg.dataSize);
+
+                        if (pending_topics_.find(topic.type_name) == pending_topics_.end())
+                        {
+                            pending_topics_[topic.type_name].insert(topic.m_topic_name);
+                            mcap_size_tracker_.check_and_increase_mcap_size_final(mcap_size_tracker_.get_blank_schema_size(configuration_.ros2_types, topic.type_name));
+                            mcap_size_tracker_.check_and_increase_mcap_size_final(mcap_size_tracker_.get_blank_channel_size(topic.m_topic_name));
+                        }
+                        else if (!pending_topics_[topic.type_name].count(topic.m_topic_name))
+                        {
+                            pending_topics_[topic.type_name].insert(topic.m_topic_name);
+                            mcap_size_tracker_.check_and_increase_mcap_size_final(mcap_size_tracker_.get_blank_channel_size(topic.m_topic_name));
+                        }
+
+                        add_data_nts_(msg, topic);
+                    }
                 }
                 else
                 {
-                    // No schema available + no pending samples -> Add to buffer with blank schema
-                    add_data_nts_(msg, topic);
+                    logInfo(
+                        DDSRECORDER_MCAP_HANDLER,
+                        "Schema for topic " << topic << " not yet available, inserting to pending samples queue.");
+
+                    mcap_size_tracker_.check_and_increase_mcap_size(msg.dataSize);
```

Comment:
Perhaps we can avoid code repetition if the same is being done for several state cases.